modules/statistics/PrecallTest.js

```diff
@@ -0,0 +1,126 @@
+        ScriptUtil.loadScript(
+            CALLSTATS_SCRIPT_URL,
+            /* async */ true,
+            /* prepend */ true,
+            /* relativeURL */ undefined,
+            /* loadCallback */ resolve);
+    });
+}
+
+/**
+ * Initializes the callstats lib and registers a callback to be invoked
+ * when there are 'preCallTestResults'.
+ *
+ * @typedef PrecallTestOptions
+ * @type {Object}
+ * @property {string} callStatsID - Callstats credentials - the id.
+ * @property {string} callStatsSecret - Callstats credentials - the secret.
+ * @property {string} statisticsId - The user name to use when initializing callstats.
+ * @property {string} statisticsDisplayName - The user display name.
+ *
+ * @param { PrecallTestOptions} options - The init options.
+ * @returns {Promise<void>}
+ */
+function _initialize(options) {
+    return new Promise((resolve, reject) => {
+        if (!options.disableThirdPartyRequests) {
+            const appId = options.callStatsID;
+            const appSecret = options.callStatsSecret;
+            const userId = options.statisticsId || options.statisticsDisplayName || Settings.callStatsUserName;
+
+            api.initialize(appId, appSecret, userId, (status, message) => {
+                if (status === 'success') {
+                    api.on(PRECALL_TEST_RESULTS, (type, result) => {
+                        if (type === 'success') {
+                            emitter.emit(PRECALL_TEST_RESULTS, result);
+                        }
+                    });
+                    _initialized = true;
+                    resolve();
+                } else {
+                    reject({
+                        status,
+                        message
+                    });
+                }
+            });
+        }
+    });
+}
+
+
+/**
+ * Loads the callstats script and initializes the library.
+ *
+ * @param {Function} onResult - The callback to be invoked when results are received.
+ * @returns {Promise<void>}
+ */
+export async function init(options) {
+    if (_initialized) {
+        throw new Error('Precall Test already initialized');
+    }
+
+    await _loadScript();
+    // eslint-disable-next-line new-cap
+    api = new window.callstats();
+
+    return _initialize(options);
+}
+
+/**
+ * Executes a pre call test.
+ *
+ * @typedef PrecallTestResults
+ * @type {Object}
+ * @property mediaConnectivity {boolean} - If there is media connectivity or not.
+ * @property throughput {number} - The average throughput.
+ * @property fractionalLoss - The packet loss.
+ * @property rtt {number} - The round trip time.
+ * @property provider {string} - It is usually 'callstats'.
+ *
+ * @returns {Promise<{PrecallTestResults}>}
+ */
+export function execute() {
+    if (!_initialized) {
+        console.log('Precall Test not initialized');
+
+        return;
+    }
+
+    return new Promise(resolve => {
+        emitter.once(PRECALL_TEST_RESULTS, result => {
+            resolve(result);
```

Comment:
how about reject when `!success` at line 56? maybe the status can be processed here

Reply:
That's what I though in the first place. 💡

The thing is: every time I call `makePrecallTest` the first result on `preCallTestResults` is a failure with a `null` payload. Then, after a while, I receive a successful result with a proper payload. But I don't know if this behaviour is consistent or will suffer changes in time.

Reply:
Oh so according to the docs[1] precall test runs continuously and I don't see any docs for the `makePrecallTest` method. Can you check with the CallStats folks about this?

[1]: https://docs.callstats.io/javascript/#callstats-startprecalltests

Reply:
I just have ignored the first failure (for now)...